### PR TITLE
Allow NULL shapes to be returned in WFS GetFeature requests

### PR DESCRIPTION
**Pull request by geographika**
When exporting data using WFS I was getting occasional unhelpful error messages `<ows:Exception exceptionCode="NoApplicableCode" locator="mapserv"/>` depending on how many records were returned. 

On further investigation, it was caused by NULL geometries. Records were returned via the query (in this case via MSSQL Server), and the shapes given the geometry type of the layer. Later in the `msProjectShapeEx` function the `msFreeShape` was called on the shapes (as there were no line parts), removing all attributes. There is however no check to see the return value of this function, so the empty shape is still added to the result cache.

When trying to write the WFS back out (using an OGR OUTPUTFORMAT) accessing the freed shape causes a memory exception ([here](https://github.com/MapServer/MapServer/blob/58512a6d624f2c4e10350231b70ed0b9058f3930/mapogroutput.cpp#L1121). Finally, they are discarded when attempting to [reproject again](https://github.com/MapServer/MapServer/blob/58512a6d624f2c4e10350231b70ed0b9058f3930/mapogroutput.cpp#L1145) and the generic error is returned to the WFS client. 

This pull request avoids freeing a shape with a null geometry, and instead sets its geometry type to NULL (so it can avoid reprojection later) and allows it to be returned as part of a WFS result set. See also related issue #5192. 

If this is not desired behaviour, then at least a check should be made on the return value of `msProjectShapeEx` in `msquery.c` and the shape discarded, or an error returned. 

**Update** - note the errors only occur if the data is requested in a different projection than the layer projection.

### CODE REVIEW

**Comment by jmckenna** [comment]
thanks @geographika  Can you also include an msautotest for this case?

**Comment by jratike80** [comment]
I am remembering that in the OGC standards when a spatial filter is used the features with null geometries should be included in the result set. Maybe that should be tested also - after reading the standards to confirm what is the right behavior.

**Comment by geographika** [comment]
> thanks @geographika Can you also include an msautotest for this case?

The patch works fine for the MSSQL driver, however in trying to make a test case with a Shapefile NULLs are always eliminated when reading the Shapefile, which means they can never appear in the result set. 

> I am remembering that in the OGC standards when a spatial filter is used the features with null geometries should be included in the result set. Maybe that should be tested also - after reading the standards to confirm what is the right behavior.

Ideally, all drivers would be consistent in regard to the behaviour. There is a risk each driver handles reading NULL features differently, which could require lots of changes/tests. 

Further testing has revealed another bug in MapServer GetFeature output when feature caching is used - features are reprojected prior to adding to the cache, and then reprojected again when outputting to OGR leading to cached features appearing in a completely different location to the rest of the output.

**Comment by jratike80** [comment]
At least in OGC API Features bbox must return also the features which do not have a geometry https://github.com/opengeospatial/ogcapi-features/issues/765.

File: mapquery.c [hunk 1]
```diff
@@ -1244,6 +1244,7 @@ int msQueryByRect(mapObj *map)
               break;
             }
         }
+        // msProjectShapeEx can free the shape if there is a NULL geometry, but there is no check on the return status of this function
         msProjectShapeEx(reprojector, &shape);
```
[the remark this answers is not in the document]
Comment (geographika) [reply mapquery.c]: Thanks @rouault - I've added this in.

**Comment by geographika** [comment]
Test added with a NULL feature, reprojection, and OGR export. 

Note this pull request does not change which MapServer inputs allow NULL values. The MSSQL driver allows NULL shapes, but for example the Shapefile reader [filters these out](https://github.com/MapServer/MapServer/blob/27d02c5b6c80b9671b59ba8b988b08a6708ec1c6/mapshape.c#L2420).

**Comment by geographika** [comment]
@jmckenna - as this could change the number of features returned, and include null geometries, it could risk breaking existing apps. Maybe best to avoid backporting and add to the 8.2 release only?

**Comment by rouault** [comment]
Merging. I think it's fine for backporting

**Pull request merged**